### PR TITLE
Feature/apps 1591 create study modal

### PR DESCRIPTION
**Pull request by anzioka**
Created an ActionModal for creating a new study. Functionality to create study will come later.

create_study

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
     "immutable": "4.0.0-rc.9",
     "lattice": "0.71.1",
     "lattice-auth": "0.16.0",
+    "lattice-fabricate": "^0.7.2",
```
Comment (UnsungHero97) [inline package.json]: Please use `~` instead of `^`.

File: src/containers/studies/StudiesActions.js [hunk 2]
```diff
@@ -8,7 +8,12 @@ import type { RequestSequence } from 'redux-reqseq';
 const GET_STUDIES :'GET_STUDIES' = 'GET_STUDIES';
 const getStudies :RequestSequence = newRequestSequence(GET_STUDIES);
 
+const CREATE_STUDY :'CREATE_STUDY' = 'CREATE_STUDY';
+const createStudy :RequestSequence = newRequestSequence(CREATE_STUDY);
+
 export {
   GET_STUDIES,
-  getStudies
+  CREATE_STUDY,
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesActions.js:15]: Please sort.

File: src/containers/studies/StudiesContainer.js [hunk 3]
```diff
@@ -62,17 +63,36 @@ type Props = {
   studies :List;
 };
 
-class StudiesContainer extends Component<Props> {
+type State = {
+  isCreateStudyModalVisible :boolean;
+}
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesContainer.js]: You need a semi-colon here.

```
type State = {
  ...
}; <- semicolon
```

File: src/containers/studies/StudiesReducer.js [hunk 4]
```diff
@@ -8,11 +8,13 @@ import type { SequenceAction } from 'redux-reqseq';
 
 import {
   GET_STUDIES,
+  CREATE_STUDY,
   getStudies
 } from './StudiesActions';
 
 const INITIAL_STATE :Map<*, *> = fromJS({
   [GET_STUDIES]: { requestState: RequestStates.STANDBY },
+  [CREATE_STUDY]: { requestState: RequestStates.STANDBY },
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesReducer.js:17]: Please sort.

File: src/containers/studies/components/CreateStudyForm.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  version: {
+    classNames: 'column-span-12'
+  }
+};
+
+type Props = {
+  formData :Object;
+  handleOnChange :Function;
+  isSubmitting :boolean;
+}
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyForm.js]: Missing semicolon.

File: src/containers/studies/components/CreateStudyForm.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    },
+    version: {
+      title: 'Study Version',
+      type: 'string'
+    },
+    email: {
+      title: 'Contact Email',
+      type: 'string'
+    }
+  },
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyForm.js]: You should wrap this in another layer of `properties` because it'll render nicer in the modal. I'll show you.

File: src/containers/studies/components/CreateStudyModal.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  data :Object;
+  isSubmitting :boolean;
+}
+
+
+const ModalBodyWrapper = styled.div`
+  min-width: 440px;
+`;
+
+const initialFormDataState :Object = {
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: If you're pulling out initial state into a constant, why is it not being used anywhere?

File: src/containers/studies/components/CreateStudyModal.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  email: '',
+  group: '',
+  name: '',
+  version: '',
+};
+class CreateStudyModal extends Component<Props, State> {
+  constructor(props :Props) {
+    super(props);
+    this.state = {
+      data: { ...initialFormDataState },
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: You don't need to spread. `data: initialFormDataState` is identical. Also, I recommend removing `initialFormDataState` and just setting the state here, unless you plan on reusing `initialFormDataState` somewhere else.

File: src/containers/studies/components/CreateStudyModal.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  isVisible :boolean;
+  requestStates:{
+    CREATE_STUDY :RequestState
+  }
+};
+
+type State = {
+  data :Object;
+  isSubmitting :boolean;
+}
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: Missing semicolon.

File: src/containers/studies/components/CreateStudyModal.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  handleOnCloseModal :Function;
+  isVisible :boolean;
+  requestStates:{
+    CREATE_STUDY :RequestState
+  }
+};
+
+type State = {
+  data :Object;
+  isSubmitting :boolean;
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: We've moved away from using booleans to indicate request state. Please use a proper `RequestState`, which is probably `CREATE_STUDY`.

Separately, from what I can see, `isSubmitting` isn't real. It is not passed in from a parent component. Please double check this prop.

File: src/containers/studies/components/CreateStudyModal.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+import CreateStudyForm from './CreateStudyForm';
+
+import { CREATE_STUDY } from '../StudiesActions';
+
+type Props = {
+  handleOnCloseModal :Function;
+  isVisible :boolean;
+  requestStates:{
+    CREATE_STUDY :RequestState
+  }
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: Missing semicolon.

File: src/containers/studies/components/CreateStudyModal.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+import { connect } from 'react-redux';
+import { RequestStates } from 'redux-reqseq';
+import type { RequestState } from 'redux-reqseq';
+
+import CreateStudyForm from './CreateStudyForm';
+
+import { CREATE_STUDY } from '../StudiesActions';
+
+type Props = {
+  handleOnCloseModal :Function;
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: The `:Function` type should be avoided. The annotation for this should be `() => void;`.

https://flow.org/en/docs/types/functions/#toc-function-type

File: src/containers/studies/components/CreateStudyModal.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+
+type State = {
+  data :Object;
+  isSubmitting :boolean;
+}
+
+
+const ModalBodyWrapper = styled.div`
+  min-width: 440px;
+`;
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js:30]: No change required.

This probably belongs in `src/components/modals`, similar to how `lattice-orgs` has it.

File: src/containers/studies/components/CreateStudyModal.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        <ModalBodyWrapper>
+          <CreateStudyForm
+              formData={data}
+              handleOnChange={this.handleOnChange}
+              isSubmitting={isSubmitting} />
+        </ModalBodyWrapper>
+      ),
+      [RequestStates.FAILURE]: (
+        <ModalBodyWrapper>
+          <span> Failed to create a new study. Please try again </span>
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: You need a period at the end of `Please try again.`.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/studies/components/CreateStudyForm.js (hunk 6 above)
Comment (solomontang) [reply src/containers/studies/components/CreateStudyForm.js]: You should also the data processing utils to handle your data. Unfortunately the documentation isn't very good but there is a working example of how to set up the schema and such on storybook

https://openlattice.github.io/lattice-fabricate/?path=/story/form--data-processing-w-edits-delete

File: src/containers/studies/components/CreateStudyModal.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    this.setState({
+      data: {
+        description: formData.description || '',
+        email: formData.email || '',
+        group: formData.group || '',
+        name: formData.name || '',
+        version: formData.version || ''
+      }
+    });
+  }
```
Comment (solomontang) [inline src/containers/studies/components/CreateStudyModal.js]: You only need to supply an `onChange` if you want to execute side effects (this that occur outside of updating form field states). Since this callback function is only controlling the state which is already managed by the Form component internally, all this will do is incur extraneous render cycles.

You can track this by dropping a `console.log` in the `render` method of `CreateStudyForm` and you should see that it is triggered on every field update. By omitting this function, you should still have the same behavior without re-rendering the entire parent component.

File: src/containers/studies/components/CreateStudyForm.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+const CreateStudyForm = ({ formData, handleOnChange, isSubmitting } :Props) => (
+  <Form
+      formData={formData}
+      hideSubmit
+      isSubmitting={isSubmitting}
+      onChange={handleOnChange}
+      schema={dataSchema}
+      uiSchema={uiSchema} />
```
Comment (solomontang) [inline src/containers/studies/components/CreateStudyForm.js]: Fabricate is built on top of the `react-jsonschema-form` library and for the basics, you can use their website for documention. Our internal widgets try to match what is provided out of the box from this library. However, you're probably going to need a rundown of how the `DataProcessingUtils` work.

You do not need to provide an `onChange` prop to the `Form` if you are not executing any side effects.
https://react-jsonschema-form.readthedocs.io/en/latest/#form-initialization

Please provide an onSubmit
https://react-jsonschema-form.readthedocs.io/en/latest/#form-initialization

I suspect CreateStudyForm will not need to prepopulate any data and thus the `formData` prop can also be omitted.
https://react-jsonschema-form.readthedocs.io/en/latest/#form-initialization

**Comment by anzioka** [comment]
modal

Added functionality to create a new study by leveraging fabricate's DataProcessingUtils. The shiny new study form includes input validation (name and email are required fields).

**Comment by anzioka** [comment]
Screen Shot 2019-12-11 at 5 44 08 PM

File: src/containers/studies/StudiesContainer.js [hunk 17]
```diff
@@ -55,24 +57,50 @@ const CenterText = styled.div`
...
-    // TODO: open modal to create study
+    this.setState({
+      isCreateStudyModalVisible: true
+    });
+  }
+  handleOnCloseModal = () => {
+    const { actions } = this.props;
+
+    // this ensures that the modal always renders CreateStudyForm when opened
+    actions.resetRequestState(CREATE_STUDY);
```
Comment (solomontang) [inline src/containers/studies/StudiesContainer.js]: Judging from when this action is dispatched, I suspect you might have a visual bug.

As the form is fading out, does the modal view also reset back to its standby state with an empty form instead of the success message? This is probably happening since the state is reset before the animation can finish. Ideally, you would couple the `resetRequestState` when the contents of the Modal are unmounted.

File: src/containers/studies/components/CreateStudyForm.js [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+      [getPageSectionKey(1, 1), getEntityAddressKey(0, CHRONICLE_STUDIES, STUDY_ID)], uuid());
+
+    const entityData = processEntityData(newFormData, allEntitySetIds, propertyTypeIds);
+    const assocationEntityData = {};
+
+    dispatch(createStudy({
+      entityData,
+      assocationEntityData
+    }));
+  });
```
Comment (solomontang) [inline src/containers/studies/components/CreateStudyForm.js]: Yay hooks!

This call to `useCallback` does not have a dependency list. This means that the callback is not really memoized since it will change upon every render. I recommend removing `useCallback` until you actually observe performance consequences or bad render cycles.

I suspect you should have an eslint error if you have installed the necessary plugins. You should do your best to heed the linter when doing anything Hook related as it will stop an entire class of bugs ahead of time.

References:
https://reactjs.org/docs/hooks-reference.html#usecallback
https://www.npmjs.com/package/eslint-plugin-react-hooks

File: src/containers/studies/components/CreateStudyModal.js [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+const CreateStudyModal = (props :Props) => {
+  const formRef = useRef();
+
+  const { isVisible, handleOnCloseModal, requestStates } = props;
+
+  const handleOnSubmit = useCallback(() => {
+    if (formRef.current) {
+      formRef.current.submit();
+    }
+  });
```
Comment (solomontang) [inline src/containers/studies/components/CreateStudyModal.js]: Same as the previous comment about `useCallback`. Either provide the exhaustive dependency array, or don't use `useCallback`.

**Review by solomontang — CHANGES_REQUESTED** [review]

File: src/containers/studies/components/CreateStudyForm.js (hunk 18 above)
Comment (solomontang) [reply src/containers/studies/components/CreateStudyForm.js]: As a follow-up clarification, lifecycle methods are concerned with when some change should occur. `componentDidMount`, `componentDidUpdate`, `componentWillUnmount` are event driven. Often times, this leads to repeated code across different lifecycle methods and manual checks for equality/change.

Hooks on the other hand are concerned with state synchronization. When you provide a dependency list, you're declaring that your effect/memo/callback is in sync with those specific items by reference equality. By not providing a dependency list, you're saying that your effect/memo/callback is in sync with everything and will update on every render. Similarly, providing an empty dependency list means your hook synchronizes with nothing and will only be run once.

In short, using hooks can be beneficial and fun, but it does take changing the way you have previously approached React.

File: src/containers/studies/components/CreateStudyModal.js [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+        </ModalBodyWrapper>
+      ),
+      [RequestStates.SUCCESS]: (
+        <ModalBodyWrapper>
+          <span> Successfully created a new study. </span>
+        </ModalBodyWrapper>
+      )
+    };
+    return components;
+  };
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: This doesn't need to be a function, `requestStateComponents` can be a plain object.

File: src/containers/studies/components/CreateStudyModal.js [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+type Props = {
+  handleOnCloseModal :() => void;
+  isVisible :boolean;
+  actions:{
+    createStudy :RequestSequence
+  };
+  requestStates:{
+    CREATE_STUDY :RequestState
+  };
+};
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: Please sort the props alphabetically.

File: src/containers/studies/components/CreateStudyModal.js [hunk 22]
```diff
@@ -0,0 +1,94 @@
+/*
+ * @flow
+ */
+import React, { useCallback, useRef } from 'react';
+
+import styled from 'styled-components';
+import { Map } from 'immutable';
+import {
+  ActionModal
+} from 'lattice-ui-kit';
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: You only need to split into multiple lines if there's 4 or more imports.

File: src/core/edm/EDMReducer.js [hunk 23]
```diff
@@ -112,13 +117,15 @@ export default function edmReducer(state :Map<*, *> = INITIAL_STATE, action :Obj
             .set('entityTypes', entityTypes.asImmutable())
             .set('entityTypesIndexMap', entityTypesIndexMap.asImmutable())
             .set('propertyTypes', propertyTypes.asImmutable())
+            .set('propertyTypesFqnIdMap', propertyTypesFqnIdMap.asImmutable())
```
Comment (UnsungHero97) [inline src/core/edm/EDMReducer.js:122]: You haven't defined this in `INITIAL_STATE`.

File: src/core/edm/EDMReducer.js [hunk 24]
```diff
@@ -98,6 +100,9 @@ export default function edmReducer(state :Map<*, *> = INITIAL_STATE, action :Obj
                 .setTitle(pt.title)
                 .setType(pt.type)
                 .build();
+
+              const typeFqn :FullyQualifiedName = new FullyQualifiedName(propertyType.type);
```
Comment (UnsungHero97) [inline src/core/edm/EDMReducer.js]: Use the shorter `FQN` annotation:

```
import type { FQN } from 'lattice';
const fqn :FQN = new FullyQualifiedName(propertyType.type);
```

File: src/containers/studies/components/CreateStudyForm.js [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+  const propertyTypeIds :Map = useSelector((store :Map) => store.getIn(['edm', 'propertyTypesFqnIdMap']));
+  const dispatch = useDispatch();
+
+  const handleSubmit = useCallback((payload :any) => {
+    let { formData: newFormData } = payload;
+    newFormData = setIn(newFormData,
+      [getPageSectionKey(1, 1), getEntityAddressKey(0, CHRONICLE_STUDIES, STUDY_ID)], uuid());
+
+    const entityData = processEntityData(newFormData, allEntitySetIds, propertyTypeIds);
+    const assocationEntityData = {};
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyForm.js]: This is misspelled; should be `associationEntityData`.

File: src/core/sagas/data/DataSagas.js [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+  if (value === null || value === undefined) {
+    workerResponse.error = ERR_ACTION_VALUE_NOT_DEFINED;
+    yield put(submitDataGraph.failure(id, workerResponse.error));
+    return workerResponse;
+  }
+  try {
+    yield put(submitDataGraph.request(id, value));
+
+    const dataGraph = (new DataGraphBuilder())
+      .setAssociations(value.assocationEntityData)
```
Comment (UnsungHero97) [inline src/core/sagas/data/DataSagas.js]: This is misspelled; should be `associationEntityData`.

File: src/containers/studies/components/CreateStudyForm.js [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+    dispatch(createStudy({
+      entityData,
+      assocationEntityData
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyForm.js]: This is misspelled; should be `associationEntityData`.

File: src/containers/studies/components/CreateStudyForm.js [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+const CreateStudyForm = (props :Props, ref) => {
+
+  const allEntitySetIds :Map = useSelector((store :Map) => store.getIn(['edm', 'entitySetIds']));
+  const propertyTypeIds :Map = useSelector((store :Map) => store.getIn(['edm', 'propertyTypesFqnIdMap']));
+  const dispatch = useDispatch();
+
+  const handleSubmit = useCallback((payload :any) => {
+    let { formData: newFormData } = payload;
+    newFormData = setIn(newFormData,
+      [getPageSectionKey(1, 1), getEntityAddressKey(0, CHRONICLE_STUDIES, STUDY_ID)], uuid());
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyForm.js]: Please fix formatting:

```
newFormData = setIn(
  newFormData,
  [getPageSectionKey(1, 1), getEntityAddressKey(0, CHRONICLE_STUDIES, STUDY_ID)],
  uuid(),
);
```

File: src/containers/studies/components/CreateStudyForm.js [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+import { ENTITY_SET_NAMES } from '../../../core/edm/constants/EntitySetNames';
+import { PROPERTY_TYPE_FQNS } from '../../../core/edm/constants/FullyQualifiedNames';
+import { createStudy } from '../StudiesActions';
+
+const { STUDY_ID } = PROPERTY_TYPE_FQNS;
+const { CHRONICLE_STUDIES } = ENTITY_SET_NAMES;
+const { processEntityData, getEntityAddressKey, getPageSectionKey } = DataProcessingUtils;
+
+type Props = {
+};
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyForm.js]: Why is this empty?

File: src/containers/studies/StudiesContainer.js [hunk 30]
```diff
@@ -109,21 +137,26 @@ class StudiesContainer extends Component<Props> {
...
       </>
     );
   }
 }
 const mapStateToProps = (state :Map) => ({
   requestStates: {
-    [GET_STUDIES]: state.getIn(['studies', GET_STUDIES, 'requestState'])
+    [GET_STUDIES]: state.getIn(['studies', GET_STUDIES, 'requestState']),
+    [CREATE_STUDY]: state.getIn(['studies', CREATE_STUDY, 'requestState']),
+
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesContainer.js]: Please remove empty line.

File: src/containers/studies/StudiesReducer.js [hunk 31]
```diff
@@ -31,6 +42,14 @@ export default function studiesReducer(state :Map<*, *> = INITIAL_STATE, action
       });
     }
 
+    case createStudy.case(action.type): {
+      return createStudy.reducer(state, action, {
+        REQUEST: () => state.setIn([CREATE_STUDY, 'requestState'], RequestStates.PENDING),
+        SUCCESS: () => state.setIn([CREATE_STUDY, 'requestState'], RequestStates.SUCCESS),
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesReducer.js]: On success, you need to update state with the new study so that it shows up in the list.

File: src/containers/studies/StudiesSagas.js [hunk 32]
```diff
@@ -50,7 +56,36 @@ function* getStudiesWatcher() :Generator<*, *, *> {
   yield takeEvery(GET_STUDIES, getStudiesWorker);
 }
 
+
+function* createStudyWorker(action :SequenceAction) :Generator<*, *, *> {
+  const { id, value } = action;
+  let response = {};
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: This doesn't need to be a `let`. It can be a `const` on line 66.

File: src/containers/studies/StudiesContainer.js (hunk 17 above)
Comment (anzioka) [reply src/containers/studies/StudiesContainer.js]: You are right! There was a visual bug, though almost unnoticeable. The success message was rendered but on closing the modal an empty form would be temporarily rendered before the modal closed. I fixed the issue by calling resetRequestState before the modal was rendered (when the isVisible prop is set to true).

File: src/containers/studies/StudiesSagas.js [hunk 33]
```diff
@@ -50,7 +56,36 @@ function* getStudiesWatcher() :Generator<*, *, *> {
...
+
+    response = yield call(submitDataGraphWorker, submitDataGraph(value));
+    if (response.error) {
+      throw response.error;
+    }
+
+    yield put(createStudy.success(id));
+  }
+  catch (error) {
+    LOG.error('caught exception in createStudy()', error);
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: This should be `LOG.error(action.type)`. Please follow the patterns consistently.

File: src/containers/studies/components/CreateStudyModal.js [hunk 34]
```diff
@@ -0,0 +1,94 @@
...
+import type { RequestSequence, RequestState } from 'redux-reqseq';
+
+import CreateStudyForm from './CreateStudyForm';
+
+import { CREATE_STUDY, createStudy } from '../StudiesActions';
+
+type Props = {
+  handleOnCloseModal :() => void;
+  isVisible :boolean;
+  actions:{
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: There should be a space before the colon... `actions :{`

File: src/containers/studies/components/CreateStudyModal.js [hunk 35]
```diff
@@ -0,0 +1,94 @@
...
+
+const mapStateToProps = (state :Map) => ({
+  requestStates: {
+    [CREATE_STUDY]: state.getIn(['studies', CREATE_STUDY, 'requestState'])
+  },
+});
+
+const mapDispatchToProps = (dispatch :Function) => ({
+  actions: bindActionCreators({
+    createStudy,
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: This action is not used in the component.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/studies/components/CreateStudyModal.js [hunk 36]
```diff
@@ -16,16 +14,15 @@ import type { RequestSequence, RequestState } from 'redux-reqseq';
...
-import { CREATE_STUDY, createStudy } from '../StudiesActions';
+import { CREATE_STUDY } from '../StudiesActions';
 
 type Props = {
-  handleOnCloseModal :() => void;
-  isVisible :boolean;
-  actions:{
-    createStudy :RequestSequence,
+  actions :{
     resetRequestState :RequestSequence
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js:19]: `resetRequestState` is also not used anywhere. Please remove.

File: src/containers/studies/components/CreateStudyModal.js [hunk 37]
```diff
@@ -93,7 +87,6 @@ const mapStateToProps = (state :Map) => ({
 
 const mapDispatchToProps = (dispatch :Function) => ({
   actions: bindActionCreators({
-    createStudy,
     resetRequestState
   }, dispatch)
 });
```
Comment (UnsungHero97) [inline src/containers/studies/components/CreateStudyModal.js]: Given that the component doesn't use any actions, it's safe to remove `mapDispatchToProps` entirely.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: package.json [hunk 38]
```diff
@@ -77,6 +79,7 @@
     "eslint-plugin-jest": "~23.1.0",
     "eslint-plugin-jsx-a11y": "~6.2.0",
     "eslint-plugin-react": "~7.17.0",
+    "eslint-plugin-react-hooks": "^2.3.0",
```
Comment (UnsungHero97) [inline package.json]: Please use `~` instead of `^`.

Comment (solomontang) [inline package.json]: Pin the version to patch by using `~` instead of `^`

File: src/containers/studies/StudiesContainer.js [hunk 39]
```diff
@@ -55,24 +57,51 @@ const CenterText = styled.div`
...
   componentDidMount() {
     const { actions } = this.props;
     actions.getStudies();
   }
+
   openCreateStudyModal = () => {
-    // TODO: open modal to create study
+    const { actions } = this.props;
+    // necessary after a successful or failed CREATE_STUDY action
+    actions.resetRequestState(CREATE_STUDY);
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesContainer.js:89]: Why do you trigger this action before the modal opens?

File: src/containers/studies/StudiesContainer.js [hunk 40]
```diff
@@ -109,21 +138,26 @@ class StudiesContainer extends Component<Props> {
...
+
   },
   studies: state.getIn(['studies', 'studies'], List())
 });
 
 const mapDispatchToProps = (dispatch :Function) => ({
   actions: bindActionCreators({
     getStudies,
+    resetRequestState,
     goToRoute: RoutingActions.goToRoute,
```
Comment (solomontang) [inline src/containers/studies/StudiesContainer.js]: You should get in the habit of sorting. We're probably going to be using prettier soon so this will be less of an ordeal in the future.

File: src/containers/studies/StudiesReducer.js [hunk 41]
```diff
@@ -26,11 +37,26 @@ export default function studiesReducer(state :Map<*, *> = INITIAL_STATE, action
...
       });
     }
 
+    case createStudy.case(action.type): {
+      return createStudy.reducer(state, action, {
+        REQUEST: () => state.setIn([CREATE_STUDY, 'requestState'], RequestStates.PENDING),
+        FAILURE: () => state.setIn([CREATE_STUDY, 'requestState'], RequestStates.FAILURE),
+        SUCCESS: () => {
+          const { studyUUID } = action.value;
+          const study = state.getIn(['studies', studyUUID], Map());
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesReducer.js]: I'd be surprised if this is actually working. In the worker, you have:

```
yield put(createStudy.success(id));
```

This means `action.value` is equal to `{}` and `studyUUID` will be undefined.

Separately, why are you getting the new study from state? It's a new study, which means it won't be in state, which means you'll always be storing an empty map below.

File: src/containers/studies/StudiesSagas.js [hunk 42]
```diff
@@ -50,7 +56,35 @@ function* getStudiesWatcher() :Generator<*, *, *> {
   yield takeEvery(GET_STUDIES, getStudiesWorker);
 }
 
+
+function* createStudyWorker(action :SequenceAction) :Generator<*, *, *> {
+  const { id, value } = action;
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js]: Please move this inside the `try` block.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/containers/studies/StudiesContainer.js (hunk 39 above)
Comment (solomontang) [reply src/containers/studies/StudiesContainer.js:89]: I suspect this is to clear any previous lingering request states. Before, he was clearing on close but since it was being called in the parent container, it was causing flickering when then modal was transitioning out. 

Ideally, you would co-locate as much of the logic relating to the contents of your modal all in one place. The Modal will unmount its children when the transition is complete. So as long as the contents of the modal are responsible for their own cleanup, there won't be any issues.

But this seems to be the next best way.

File: src/containers/studies/StudiesContainer.js (hunk 40 above)
Comment (anzioka) [reply src/containers/studies/StudiesContainer.js]: Ok, sorted.

File: src/containers/studies/StudiesReducer.js (hunk 41 above)
Comment (anzioka) [reply src/containers/studies/StudiesReducer.js]: I fixed this.

File: src/containers/studies/StudiesContainer.js (hunk 39 above)
Comment (anzioka) [reply src/containers/studies/StudiesContainer.js:89]: You are right @solomontang. After a SUCCESS/ FAILURE success state, opening the modal a second time would not render the form component, that's why I reset the state before the modal opens.

**Comment by anzioka** [comment]
Merged  feature/APPS-1592-create-study-details-page with this branch.
